x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportExplainDataFrameAnalyticsAction.java

```diff
@@ -206,31 +206,34 @@ private void estimateMemoryUsage(Task task,
     }
 
     /**
-     * Finds the first available ML node in the cluster and redirects the request to this node.
+     * Finds the a suitable node in the cluster we can run on, and redirects the request to this node.
      */
-    private void redirectToMlNode(Task task,
-                                  PutDataFrameAnalyticsAction.Request request,
-                                  ActionListener<ExplainDataFrameAnalyticsAction.Response> listener) {
-        Optional<DiscoveryNode> node = findMlNode(clusterService.state());
+    private void redirectToSuitableNode(PutDataFrameAnalyticsAction.Request request,
+                                        ActionListener<ExplainDataFrameAnalyticsAction.Response> listener) {
+        Optional<DiscoveryNode> node = findSuitableNode(clusterService.state());
         if (node.isPresent()) {
             transportService.sendRequest(node.get(), actionName, request,
                 new ActionListenerResponseHandler<>(listener, ExplainDataFrameAnalyticsAction.Response::new));
-        } else if (numLazyMLNodes > 0 || request.getConfig().isAllowLazyStart()) {
-            explain(task, request, false, listener);
         } else {
-            listener.onFailure(ExceptionsHelper.badRequestException("No ML node to run on"));
+            listener.onFailure(ExceptionsHelper.badRequestException("No ML, data or ingest node to run on"));
         }
     }
 
     /**
-     * Finds the first available ML node in the cluster state.
+     * Prefer the first available ML node in the cluster state.
+     * If there isn't one, find the first data or ingest node.
      */
-    private static Optional<DiscoveryNode> findMlNode(ClusterState clusterState) {
+    private static Optional<DiscoveryNode> findSuitableNode(ClusterState clusterState) {
         for (DiscoveryNode node : clusterState.getNodes()) {
             if (MachineLearning.isMlNode(node)) {
                 return Optional.of(node);
             }
         }
+        for (DiscoveryNode node : clusterState.getNodes()) {
+            if (node.isDataNode() || node.isIngestNode()) {
+                return Optional.of(node);
+            }
+        }
```

Comment:
Do we need to worry about the ml plugin being enabled or not?

Reply:
Yes, good point.  We demand that the ML plugin is enabled on master-eligible nodes, so I think the only safe option will be to redirect to one of these.  I will try to get it to prefer the master-eligible nodes that aren't the active master.

Reply:
OK, I think that will work. 

It would be cool if we could determine if the plugin is enabled or not given the info in the discovery node data. That way we are not restricted to only master nodes.

But, I am not sure that is possible. 

So, on a minimal deployment (just one node), this will work as we always need at least one master node :).

Reply:
The only times a redirection is needed now are:

1. When the connection comes from a transport client (impossible in 8.x), because this code will run on the transport client node
2. When the connection comes from a node client (only possible in tests now, but a crucial case for getting CI to pass), because this code will run on the node client node
3. When the connection comes from the REST API but goes to a pure coordinating node in the cluster - we could avoid this case too in 8.x if there was an easy way to distinguish node client nodes from coordinating nodes

There is talk of maybe adding a dedicated `coordinating` node role in the future.  That would satisfy the requirement of 3.

The TL;DR is that redirections will be rarer in production than in our tests, and can probably be made even rarer in the future.  So I don't think it's a good idea to add even more attributes to the cluster state just for this.